Add optimistic locking support to dynamo backend

rosstimothy opened the pull request:
The dynamo test was also updated to allow the table name to be overridden via an environment variable.

Contributes to #30416.

espadolini commented on lib/backend/dynamo/dynamodbbk.go:
```diff
@@ -562,6 +572,47 @@ func (b *Backend) Delete(ctx context.Context, key []byte) error {
+		return trace.Wrap(err)
+	}
+	input := dynamodb.DeleteItemInput{Key: av, TableName: aws.String(b.TableName)}
+	input.SetConditionExpression("#v = :rev")
+	input.SetExpressionAttributeNames(map[string]*string{"#v": aws.String("Revision")})
```
Doesn't this work? I'm not sure why we do this in `CompareAndSwap`.

```suggestion
	input.SetConditionExpression("Revision = :rev")
```

espadolini commented on lib/backend/dynamo/dynamodbbk.go:
```diff
@@ -562,6 +572,47 @@ func (b *Backend) Delete(ctx context.Context, key []byte) error {
+// ConditionalDelete deletes item by key if the provided revision matches
+// the revision of the item in Dynamo.
+func (b *Backend) ConditionalDelete(ctx context.Context, key []byte, rev string) error {
+	if len(key) == 0 {
+		return trace.BadParameter("missing parameter key")
+	}
```
`create` (as called by `Create`, `Update` and `Put` and now `ConditionalUpdate`) doesn't check that the key is nonempty.

espadolini replied:
If anything we should allow deleting a key that's empty but not write it - right now we're in the weird position of allowing a Put with an empty key with no way to delete it 😬 

If we can distinguish between an empty but set FullPath and no FullPath (as we use that for existence), why not just allow the empty key?

espadolini commented on lib/backend/dynamo/dynamodbbk.go:
```diff
@@ -838,42 +881,55 @@ func trimPrefix(key string) []byte {
 		input.SetConditionExpression("attribute_exists(FullPath)")
 	case modePut:
+	case modeConditionalUpdate:
+		input.SetExpressionAttributeNames(map[string]*string{"#v": aws.String("Revision")})
+		input.SetExpressionAttributeValues(map[string]*dynamodb.AttributeValue{":rev": {S: aws.String(item.Revision)}})
+		input.SetConditionExpression("#v = :rev AND attribute_exists(FullPath)")
```
Same as in `ConditionalDelete`.

```suggestion
		input.SetExpressionAttributeValues(map[string]*dynamodb.AttributeValue{":rev": {S: aws.String(item.Revision)}})
		input.SetConditionExpression("Revision = :rev AND attribute_exists(FullPath)")
```

espadolini commented on lib/backend/dynamo/dynamodbbk.go:
```diff
@@ -632,22 +682,14 @@ const (
 	tableStatusMissing
 	tableStatusNeedsMigration
 	tableStatusOK
+	modeConditionalUpdate
```
Wrong `const` group

rosstimothy committed:
Add optimistic locking support to dynamo backend

The dynamo test was also updated to allow the table name to be
overridden via an environment variable.

Contributes to #30416.
